Add support for DTMF via PCMA

pali opened the pull request:
Some telecommunication operators prefer PCMA codec over PCMU. Analogue
fixed telephone lines support DTMF only via in-band channel so RFC2833
cannot be used.

This patch adds simple support for generating and extracting DTMF via PCMA
codec. PCMU is still preferred and default. DTMF over PCMA would be used
only when user explicitly configure PCMA via rtp_param or when remote peer
is using PCMA (and not PCMU).

noxxi commented:
Thanks for the patch. Could you please add a test for this new functionality to t/19_call_with_dtmf.t too?

pali commented:
Done! I added PCMA test into file t/19_call_with_dtmf.t and also added a new simple test t/24_dtmf_audio.t directly for DTMF audio module which does not use SIP.

During testing I found two issues. First issue was that I used incorrect formula for calculating a-law algorithm and second issue was in parsing media attributes in sub dtmf (name 'PCMA/8000' does not have to be present but there can be format 8 which is PCMA). Both issues I fixed and updated pull request.

If you think that that tests should be extended, please let me know. I can imagine another test: generate PCMU DTMF, convert PCMU to PCMA according to ITU-T Rec. G.711 tables and then extract DTMF from this PCMA audio. This could prove that PCMU a PCMA audio code is compatible.

noxxi commented on lib/Net/SIP/Simple/Call.pm:
```diff
@@ -452,9 +452,17 @@ sub dtmf {
+	    my $sdp_peer = $self->{param}{sdp_peer};
+	    if ($sdp_peer) {
+		$type = $sdp_peer->name2int('PCMU/8000','audio') unless defined $type;
+		$type = $sdp_peer->name2int('PCMA/8000','audio') unless defined $type;
```
This line will never be executed since $type is already defined in the previous line.

pali replied:
This is not truth. `name2int` may return undef if `PCMU/8000` on previous line does not exists in `lines` part of sdp peer. This happens when SDP contains only fmt id (0) without fmt name (PCMU).

pali replied:
I changed code to use `//`.

noxxi commented on lib/Net/SIP/Simple/Call.pm:
```diff
@@ -452,9 +452,17 @@ sub dtmf {
+		$type = $sdp_peer->name2int('PCMA/8000','audio') unless defined $type;
+		if (!defined $type) {
+			my $media = $sdp_peer->get_media();
+			my ($fmt) = map { ($_->{media} eq 'audio') ? $_->{fmt} : () } @$media;
+			($type) = grep { $_ == 0 || $_ == 8 } @$fmt; # Only PCMU=0 and PCMA=8 are supported
```
Why not merge these two lines, i.e.

      ($type) =  map { $_->{media} eq 'audio' && ($_->{fmt} == 0 || $_->{fmt} == 8) ? $_->{fmt} : () } @$media;

pali replied:
Because $fmt is reference to array. So you cannot write `... eq 'audio' && ($_->{fmt} == ..`. `$_->{fmt}` is arrayref, so above code does not work. You would have to put `grep` into that `map`, something like this:

```perl
($type) = map { ($_->{media} eq 'audio') ? grep { $_ == 0 || $_ == 8 } @{$_->{fmt}} : () } @$media;
```

And I think that map and grep on separate lines is more readable.

But let me know what do you think. I can change it, if you prefer it.

noxxi commented on lib/Net/SIP/Simple/Call.pm:
```diff
@@ -452,9 +452,17 @@ sub dtmf {
-		|| 0; # default id for PCMU/8000
+	    my $sdp_peer = $self->{param}{sdp_peer};
+	    if ($sdp_peer) {
+		$type = $sdp_peer->name2int('PCMU/8000','audio') unless defined $type;
```
Since Net::SIP declares to require at least Perl 5.10 one could simply write

       $type //= $sdp_peer->name2int('PCMU/8000','audio')

pali replied:
Done!

noxxi commented on lib/Net/SIP/Simple/Call.pm:
```diff
@@ -452,9 +452,17 @@ sub dtmf {
+			($type) = grep { $_ == 0 || $_ == 8 } @$fmt; # Only PCMU=0 and PCMA=8 are supported
+		}
+	    }
+	    $type = 0 unless defined $type; # default id for PCMU/8000
```
This could again by done simpler with `$type //= 0`.

pali replied:
Done!

pali commented:
@noxxi Anything else related to this pull request is needed?

noxxi commented:
Thanks for your contribution.
I've merged it into Net::SIP but add some smaller enhancements and fixes on top, see bf9e4fc8d.

pali commented:
ok, great!
